### DISCUSSION TITLE
Tiling tensor with overalp

### DISCUSSION THREAD

**Discussion by IuriiKobein** [NVIDIA/cutlass#1836] [Q&A]
Currently I am working on validating idea to use CuTe/CUTLASS for finite difference method.
It requires to divide input by tile with overlaps to take into consideration halo region.
The input is 2D tensor that already has allocated halo regions on boundaries.

For example following 2D input divide into 3 overlapped tiles:
<pre>
In:             tile0:   tile1:   tile2:
0 0 0 0           0 0      0 0     0 0
0 1 3 0  =>       0 1 ,    1 3,    3 0
0 2 4 0           0 2      2 4     4 0
0 0 0 0           0 0      0 0     0 0
</pre>

Is there utils to do it using CuTe?

**Comment by thakkarV** [comment]
No, but @ccecka and I have discussed an idea for implementing a stencil div for example 59 in the past. This is a lower priority for us at the moment, but our internal discussions suggest that this is most certainly doable and a generalization of the im2col transformed used for convolutions.

**Reply by IuriiKobein** [reply]
For a big star stencil im2col could be inefficient. Just utilize CUDA core on a tile directly is preferable because of simplicity and precision - tf32 is not enough f64 is overkill and f32 is OK.
The question how to divide the input into overlapped tiles with minimum effort on app side and delegate as much as possible to CuTe.
From docs it looks feasible, however it's quite unclear.

**Reply by IuriiKobein** [reply]
@thakkarV I would like ask you about another approach. 
For example we have input tensor 32x48. 
It's divided by 8x8 tiler.
Now the question is there way in CuTe to take tiles (0,1), (0,2), (1,1), (1,2) and merge them into single tensor view?

**Reply by thakkarV** [reply]
You are using quite imprecise terminology. I am not sure what you mean by "merge them" here. Is it fair to say that you want each final tile to be of shape 16x16? If so, why not just use a 16x16 tiler?

**Reply by YuraCobain** [reply]
I mean having 4  tiles 16x16 that have coordinates mentioned above view on them as single 32x32 tile.

**Reply by thakkarV** [reply]
you mentioned a 8x8 tiler above, so I am confused now. can you please rephrase your question?

**Reply by YuraCobain** [reply]
Sorry, let's start from scratch.
Assume division of tensor 32x48 by tiler 16x16 into following enumerated tiles
0 2 4
1 3 5

Is possible in CuTe for example to view tiles 2, 3, 4, 5 as a single tile 32x32?

**Reply by thakkarV** [reply]
I see what you are trying to do. Here is what I would recommend you do instead: 

Think of your input tensor's layout as a surjective function from a domain of size 64x64 = 4096 onto a codomain of size 32x32 = 1024
Then tile this tensor with whatever 32x32 tiler you want.
Done.

This is exactly what example 59 does.

**Reply by IuriiKobein** [reply]
@thakkarV thanks for the recommendation. I will try this approach and be back in case of additional question.
